internal/cri/server/restart.go

```diff
@@ -100,12 +101,18 @@ func (c *criService) recover(ctx context.Context) error {
 		return fmt.Errorf("failed to list sandboxes from API: %w", err)
 	}
 	for _, sbx := range storedSandboxes {
-		if _, err := c.sandboxStore.Get(sbx.ID); err == nil {
+		sb, err := c.sandboxStore.Get(sbx.ID)
```

Comment:
Suggestion: we should merge metadata-db.sandbox-store.metadata with metadata-db.container.metadata in function `podSandboxLoader.RecoverContainer`. This loop is designed for the sandboxes which are managed by remote sandbox plugin. I think we should not involve two sources in one loop.

In the `RecoverContainer` function, we should prefer to use metadata from metadata-db.sandbox. What do you think?

-----

@dmcgowan  @mxpv 

I found that the sandboxes bucket wasn't in `v1`. Should we add the version before namespace for sandbox?

https://github.com/containerd/containerd/blob/67a0efc589af5a1a092d3489cfe2d1e9ec01506d/core/metadata/buckets.go#L309-L315

Reply:
I agree, but that seemed a big shift for this PR. This "use the underlying container for the given sandbox" vs "use the sandbox api as the source of truth to grab the sandboxes" is very confusing atm.

Reply:
> use the underlying container for the given sandbox

It's to recover sandbox created by previous release actually. 

> but that seemed a big shift for this PR

OK. What about adding a condition that if `IP` is empty, we should use metadata from sandbox bucket 😂

Reply:
I can give that a try!